aws/resource_aws_budget.go

```diff
@@ -0,0 +1,303 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/budgets"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsBudget() *schema.Resource {
+	return &schema.Resource{
+		Schema: resourceAwsBudgetSchema(),
+		Create: resourceAwsBudgetCreate,
+		Read:   resourceAwsBudgetRead,
+		Update: resourceAwsBudgetUpdate,
+		Delete: resourceAwsBudgetDelete,
+	}
+}
+
+func resourceAwsBudgetSchema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"name": {
+			Type:          schema.TypeString,
+			Optional:      true,
+			ConflictsWith: []string{"name_prefix"},
+		},
+		"name_prefix": {
+			Type:     schema.TypeString,
+			Optional: true,
+			ForceNew: true,
+		},
+		"budget_type": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		"limit_amount": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		"limit_unit": {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		"include_credit": {
```

Comment:
While its more complex right now, it feels like these should all be nested under a `cost_types` attribute to match the API. We have run into issues with other AWS services where new API attributes are introduced that are similar or the same name as existing attributes. e.g.

```go
"cost_types": {
  Type:     schema.TypeList,
  Optional: true,
  MaxItems: 1,
  Elem:     &schema.Resource{
    Schema: map[string]*schema.Schema{
      "include_credit": {
        // ...
      },
      // ...
      "use_blended": {
        // ...
      },
  },
},
```